js/privatebin.js

```diff
@@ -2761,44 +2816,42 @@ jQuery.PrivateBin = (function($, RawDeflate) {
          */
         me.setAttachment = function(attachmentData, fileName)
         {
-            // data URI format: data:[<mediaType>][;base64],<data>
+            // skip, if attachments got disabled
+            if (!$attachmentLink || !$attachmentPreview) return;
+
+            // data URI format: data:[<mimeType>][;base64],<data>
 
             // position in data URI string of where data begins
             const base64Start = attachmentData.indexOf(',') + 1;
-            // position in data URI string of where mediaType ends
-            const mediaTypeEnd = attachmentData.indexOf(';');
+            // position in data URI string of where mimeType ends
+            const mimeTypeEnd = attachmentData.indexOf(';');
 
-            // extract mediaType
-            const mediaType = attachmentData.substring(5, mediaTypeEnd);
+            // extract mimeType
+            const mimeType = attachmentData.substring(5, mimeTypeEnd);
             // extract data and convert to binary
             const rawData = attachmentData.substring(base64Start);
             const decodedData = rawData.length > 0 ? atob(rawData) : '';
 
-            // Transform into a Blob
-            const buf = new Uint8Array(decodedData.length);
-            for (let i = 0; i < decodedData.length; ++i) {
-                buf[i] = decodedData.charCodeAt(i);
-            }
-            const blob = new window.Blob([ buf ], { type: mediaType });
-
-            // Get Blob URL
-            const blobUrl = window.URL.createObjectURL(blob);
-
-            // IE does not support setting a data URI on an a element
-            // Using msSaveBlob to download
-            if (window.Blob && navigator.msSaveBlob) {
-                $attachmentLink.off('click').on('click', function () {
-                    navigator.msSaveBlob(blob, fileName);
-                });
-            } else {
-                $attachmentLink.attr('href', blobUrl);
-            }
+            let blobUrl = getBlobUrl(decodedData, mimeType);
+            $attachmentLink.attr('href', blobUrl);
 
             if (typeof fileName !== 'undefined') {
                 $attachmentLink.attr('download', fileName);
             }
 
-            me.handleBlobAttachmentPreview($attachmentPreview, blobUrl, mediaType);
+            // sanitize SVG preview
+            // prevents executing embedded scripts when CSP is not set and user
+            // right-clicks/long-taps and opens the SVG in a new tab - prevented
+            // in the preview by use of an img tag, which disables scripts, too
+            if (mimeType.match(/image\/svg/i)) {
```

Comment:
Ah this also matches [mime types like])https://stackoverflow.com/questions/11918977/right-mime-type-for-svg-images-with-fonts-embedded#11930514) `image/svg+xml`, as it is not limited to start/end of string okay?

Though maybe limit the start as there is really nothing before?

```suggestion
            if (mimeType.match(/^image\/svg/i)) {
```

Also, a unit test should test this matching, for any elemnts beginning with `image/svg` (including the one with `+xml`)

Reply:
While it is highly unlikely that future types get added to include something that ends in "image", given that they only got 3 additions over the last 40 years, it doesn't hurt to be specific on that side.

For the purposes of this logic, we must regard the mimeType to be a input from an uncontrolled source (it is set by the paste creator), so we should match it as broadly as possible for sanitization purposes. Worst case scenario is that a non-SVG gets sanitized into something that doesn't render in the preview, but that is a fail-safe and IMO preferred over the opposite case.

Hence I think we should apply your suggestion to both this pattern as well as to the ones in handleBlobAttachmentPreview() further down. And we should relax the subtype side to match any svg-ish thing, even if non-standard. I'll share an improvement for your review, shortly.

Base types are documented in:
https://www.iana.org/assignments/media-types/media-types.xhtml

Reply:
Should be addressed in [75dc346](https://github.com/PrivateBin/PrivateBin/pull/906/commits/75dc346f0fbe81c843ab0aadbfc697ded33c14bd)